### PR TITLE
MINIFICPP-1705 Upgrade and fix compilation of OpenCV

### PR DESCRIPTION
**Pull request by lordgamez**
Loading the OpenCV extension failed due to zlib library not linked to opencv-core library. Upgraded OpenCV to the latest version 4.5.5 and linked our bundled zlib to the opencv-core lib.

https://issues.apache.org/jira/browse/MINIFICPP-1705

--------------------------------------------------------------------------------------------
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Review by szaszm — APPROVED** [review]

File: cmake/BundledOpenCV.cmake [hunk 1]
```diff
@@ -138,10 +134,15 @@ function(use_bundled_opencv SOURCE_DIR BINARY_DIR)
     set_target_properties(OPENCV::libpng PROPERTIES IMPORTED_LOCATION "${OPENCV_BYPRODUCT_DIR}/${CMAKE_INSTALL_LIBDIR}/${THIRDPARTY_DIR}${PREFIX}libpng${THIRDPARTY_SUFFIX}")
     add_dependencies(OPENCV::libpng opencv-external)
 
+    add_library(OPENCV::libopenjp2 STATIC IMPORTED)
```
Comment (adamdebreceni) [inline cmake/BundledOpenCV.cmake:137]: are there some PRs incoming that verify this fix?

**Review by fgerlits — APPROVED** [review]

File: cmake/BundledOpenCV.cmake (hunk 1 above)
Comment (lordgamez) [reply cmake/BundledOpenCV.cmake:137]: OpenCV is compiled in the CI so the upgrade should be okay but I only tested the loading of the library manually. The proper way would be to have some test coverage for the OpenCV extension, but currently I'm not aware of anyone using it so it is not a priority. I created a ticket for this in Jira so we can add it in the future https://issues.apache.org/jira/browse/MINIFICPP-1709

**Review by adamdebreceni — APPROVED** [review]

File: cmake/BundledOpenCV.cmake [hunk 2]
```diff
@@ -138,10 +134,15 @@ function(use_bundled_opencv SOURCE_DIR BINARY_DIR)
...
+    set_target_properties(OPENCV::libopenjp2 PROPERTIES IMPORTED_LOCATION "${OPENCV_BYPRODUCT_DIR}/${CMAKE_INSTALL_LIBDIR}/${THIRDPARTY_DIR}${PREFIX}libopenjp2${THIRDPARTY_SUFFIX}")
```
Comment (szaszm) [inline cmake/BundledOpenCV.cmake:138]: I got this error:
```
make[2]: *** No rule to make target 'thirdparty/opencv-install/lib64/opencv4/3rdparty/liblibopenjp2.a', needed by 'bin/libminifi-opencv.so'.  Stop.
make[1]: *** [CMakeFiles/Makefile2:6131: extensions/opencv/CMakeFiles/minifi-opencv.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
```

Removing the "lib" part after `${PREFIX}` didn't help, but it changed the missing target from "liblibopenjp2.a" to "libopenjp2.a". I don't know what's the problem.

This was a clean build, only reusing FetchContent dependencies, but this uses ExternalProject, so I think it shouldn't be an issue. Any ideas?

**Review by szaszm — CHANGES_REQUESTED** [review]

File: cmake/BundledOpenCV.cmake (hunk 2 above)
Comment (lordgamez) [reply cmake/BundledOpenCV.cmake:138]: Could you check the openjp2 library on your system? Is it available under the install path and if it is what is its build path?

Comment (szaszm) [reply cmake/BundledOpenCV.cmake:138]: it's not under 'thirdparty/opencv-install/lib64/opencv4/3rdparty/liblibopenjp2.a' and it's not installed on my system elsewhere

Comment (lordgamez) [reply cmake/BundledOpenCV.cmake:138]: That looks to be right, maybe it is a ordering issue with the compilation of the targets when compiling them on multiple threads. Could you check if the issue persists if the `add_dependencies(OPENCV::libopencv-imgcodecs opencv-external)` line is expanded with the linked targets?

Comment (szaszm) [reply cmake/BundledOpenCV.cmake:138]: It's not a build order issue. Running on a single thread or repeated running on a lot of threads didn't solve the issue. Added all of the targets from `target_link_libraries` to `add_dependencies` for `imgcodecs`, but it didn't help either. For some reason, the library is not built at all, at least in my environment.

Comment (szaszm) [reply cmake/BundledOpenCV.cmake:138]: I think I just found the issue: we need to pass `-DBUILD_OPENJPEG=ON` to the opencv cmake build for it to build the library. I don't know why it worked anywhere else.

Comment (lordgamez) [reply cmake/BundledOpenCV.cmake:138]: Sorry I misread your comment, I thought you wrote that it's under the mentioned directory. It's quite strange, what OS are you building it on? Maybe I can try to reproduce it in a docker image. Also can you check if you have the same issue when building the library separately from https://github.com/opencv/opencv.git on the 4.5.5 branch?

Comment (lordgamez) [reply cmake/BundledOpenCV.cmake:138]: > I think I just found the issue: we need to pass `-DBUILD_OPENJPEG=ON` to the opencv cmake build for it to build the library. I don't know why it worked anywhere else.

Great, thanks! I'll add this to the PR

Comment (szaszm) [reply cmake/BundledOpenCV.cmake:139]: Do we actually need this new dependency though? Why did you add openjp2? I originally assumed it's just a new upstream dependency, but it seems to be just another alternative jpeg codec library.

Comment (lordgamez) [reply cmake/BundledOpenCV.cmake:139]: It just didn't compile without it, failing with linkage issues.

Comment (lordgamez) [reply cmake/BundledOpenCV.cmake]: Removed libopenjp2 dependency in e0c9e963a92e77ca72a62930f151e094851141d7

Comment (szaszm) [reply cmake/BundledOpenCV.cmake]: Summary (for the record, we continued in PM):
OpenCV changed their JPEG2000 codec from Jasper to OpenJPEG. We didn't have Jasper enabled previously, but to disable the new codec, we need to use a new option: `-DWITH_OPENJPEG=OFF`. This one defaults to ON, while the corresponding `-DBUILD_OPENJPEG` option defaulted to OFF, at least on my machine. Either including the openjp2 target (this is the openjpeg library name) or disabling `WITH_OPENJPEG` makes opencv_imgcodecs correctly link to our opencv extension.
We decided not to add JPEG2000 support for now and disable the new codec, too.

**Review by szaszm — APPROVED** [review]

**Pull request closed without merging**